### ISSUE TITLE
Semantics of EXTRACT have changed in 2.2.1

### ISSUE THREAD

**Issue by ojmason** [neo4j/neo4j#4668]
Between 2.1 and 2.2.1 the behaviour of EXTRACT has changed (breaking our system, so we cannot upgrade). This can easily be replicated by the following query:

**Setup**
`create (h:Hub {text: "Hub"})-[r1:Link]->(s1:Satellite {text: "Sat1"})`
`merge (h)-[r2:Link]->(s2:Satellite {text: "Sat2"})`

**Query**
`match (h:Hub)`
`optional match path = (h)-[r:Link]-(s:Satellite)`
`return h, extract (p in collect(path) | {`
`node: head(filter(n IN nodes(p) WHERE "Satellite" IN labels(n)))`
`})`

This creates a node with two satellite nodes hanging off it and retrieves the hub node with all the satellites attached.

![image](https://cloud.githubusercontent.com/assets/1554188/7746152/24cc2bf0-ffa8-11e4-8cf5-a1dfef322000.png)

**Behaviour on 2.1**
The _match_ query returns **a single** row, with the hub node and list containing _both_ satellite nodes.

| h | extract (p in collect(path) ! { node: head(filter(n IN nodes(p) WHERE "Satellite" IN labels(n))) }) |
| --- | --- |
| Hub | [ Sat1, Sat2 ] |

**Behaviour on 2.2.1**

The _match_ query returns **two** rows, each with the hub node and a list containing a single satellite node.

| h | extract (p in collect(path) ! { node: head(filter(n IN nodes(p) WHERE "Satellite" IN labels(n))) }) |
| --- | --- |
| Hub | [ Sat1] |
| Hub | [ Sat2] |

(In the table query the pipe character has been replaced by an exclamation mark due to table formatting)

**Comment by cleishm** [comment]
Hi @ojmason,

I've reproduced that behavior and we will look into it.

Whilst we do, I'd like to point out that you could simplify that query by writing it as follows:

```
MATCH (h:Hub)
OPTIONAL MATCH (h)-[r:Link]-(s:Satellite)
RETURN h, collect(s)
```

That also happens to return the result you're expecting in both versions.

Cheers,
Chris

**Comment by cleishm** [comment]
Indeed, we can go even simpler:

```
MATCH (h:Hub)-[r:Link]-(s:Satellite)
RETURN h, collect(s)
```

Cheers,
Chris

**Comment by ojmason** [comment]
Thanks, Chris! However, the real query we're using is a lot more complex, and I simplified it for replicating the issue here. We unfortunately cannot simplify our query in any way that works in 2.2.1.

**Comment by cleishm** [comment]
If you can post the full query, we can probably find a workaround (until such time as the underlying issue is addressed).

**Comment by ojmason** [comment]
Thanks for the offer! However, we've decided to stick with 2.1 until the issue is resolved.

**Comment by ojmason** [comment]
This seems to have been resolved in 2.2.2.

Thanks!
